eclass/rust.eclass

```diff
@@ -0,0 +1,449 @@
+			else
+				continue
+			fi
+		done
+		if [[ -z "${llvm_r1_slot}" ]]; then
+			die "${FUNCNAME}: no LLVM slot found"
+		fi
+	fi
+
+	# iterate over known slots, newest first
+	for slot in "${_RUST_SLOTS_ORDERED[@]}"; do
+		llvm_slot="${_RUST_LLVM_MAP[${slot}]}"
+		# skip higher slots
+		if [[ -n "${max_slot}" ]]; then
+			if ver_test "${slot}" -eq "${max_slot}"; then
+				max_slot=
+			elif ver_test "${slot}" -gt "${max_slot}"; then
+				continue
+			fi
+		fi
+
+		# If we're in LLVM mode we can skip any slots that don't match the selected USE
+		if [[ -n "${RUST_NEEDS_LLVM}" ]]; then
+			if [[ "${llvm_slot}" != "${llvm_r1_slot}" ]]; then
+				continue
+			fi
+		fi
+
+		if declare -f rust_check_deps >/dev/null; then
+			local RUST_SLOT="${slot}"
+			local LLVM_SLOT="${_RUST_LLVM_MAP[${slot}]}"
+			rust_check_deps && return
+		else
+			local rust_usedep="${RUST_USEDEP+[${RUST_USEDEP}]}"
+			# When checking for installed packages prefer the non `-bin` package
+			# if effort was put into building it we should use it.
+			local rust_pkgs=(
+				"dev-lang/rust:${slot}${rust_usedep}"
+				"dev-lang/rust-bin:${slot}${rust_usedep}"
+			)
+			local _pkg
+			for _pkg in "${rust_pkgs[@]}"
+				do
+					if has_version "${hv_switch}" "${_pkg}"; then
+						echo "${slot}"
+						if [[ "${_pkg}" == "dev-lang/rust:${slot}${rust_usedep}" ]]; then
+							echo source
+						else
+							echo binary
+						fi
+						return
+					fi
+				done
+		fi
+
+		# We want to process the slot before escaping the loop if we've hit the minimum slot
+		if ver_test "${slot}" -eq "${RUST_MIN_VER}"; then
+			break
+		fi
+
+	done
+
+	# max_slot should have been unset in the iteration
+	if [[ -n "${max_slot}" ]]; then
+		die "${FUNCNAME}: invalid max_slot=${max_slot}"
+	fi
+
+	die "No Rust slot${1:+ <= ${1}} satisfying the package's dependencies found installed!"
+}
+
+# @FUNCTION: get_rust_prefix
+# @USAGE: [-b|-d]
+# @DESCRIPTION:
+# Find the newest Rust install that is acceptable for the package,
+# and print an absolute path to it. If both -bin and regular Rust
+# are installed, the regular Rust is preferred.
+#
+# The options and behavior are the same as get_rust_slot.
+get_rust_prefix() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	local prefix=${BROOT}
+	[[ ${1} == -d ]] && prefix=${ESYSROOT}
+
+	local slot rust_type
+	{ read -r slot; read -r rust_type; } <<< $(get_rust_slot)
+
+	if [[ ${rust_type} == "source" ]]; then
+		echo "${prefix}/usr/lib/rust/${slot}/"
+	else
+		echo "${prefix}/opt/rust-bin-${slot}/"
+	fi
+}
+
+# @FUNCTION: rust_prepend_path
+# @USAGE: <slot> <type>
+# @DESCRIPTION:
+# Prepend the path to the specified Rust to PATH and re-export it.
+rust_prepend_path() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	[[ ${#} -ne 2 ]] && die "Usage: ${FUNCNAME} <slot> <type>"
+	local slot="${1}"
+
+	local rust_path
+	if [[ "${2}" == "source" ]]; then
+		rust_path="${BROOT}/usr/lib/rust/${slot}/bin"
+	else
+		rust_path="${BROOT}/opt/rust-bin-${slot}/bin"
+	fi
```

Comment:
It feels a bit questionable to have this logic duplicated here (in addition to `get_rust_prefix`), especially if either of these paths ever change for any reason.  It might be better to have a helper function that can do this, and then append `/bin` to the end of the result in this function.